docs/cvl/cvl2/changes.md

```diff
@@ -0,0 +1,583 @@
+`uint`.
+
+We hope that security engineers will think carefully about these changes and let
+us know of any other situations where the right thing is not simply casting to
+`mathint`.
+````
+
+```{todo}
+If you do not change this, you will see the following error:
+```
+
+(cvl2-casting)=
+### Implicit and explicit casting
+
+If every number that can be represented by one type can also be represented by
+another type, then we say that the first type is a *subtype* of the second type.
+
+For example, a `uint8` variable could have any value between `0` and `2^8-1`,
+and all of these values can be stored in a `uint16` variable, so `uint8` is a
+subtype of `uint16`.  An `int16` can also store any value between `0` and
+`2^8-1`, so `uint8` is also a subtype of `int16`.
+
+All integer types are subtypes of `mathint`, since any integer can be
+represented by a `mathint`.
+
+In CVL 1, the rules for converting between supertypes and subtypes were
+complicated; they depended not only on the types involved, but on the context
+in which the conversion happened.  CVL 2 simplifies these rules and improves the
+clarity and predictability of casts.
+
+In CVL 2, with one exception, you can always use a subtype whenever the
+supertype is accepted.  For example, you can always use a `uint8` where an
+`int16` is expected.  We say that the subtype can be "implicitly cast" to the
+supertype.
+
+The one exception is comparison operators; as mentioned above, you must add an
+explicit conversion if you want to compare two numbers with different types.
+The `to_mathint` operator exists solely for this purpose; in all other contexts
+you can simply use any number when a `mathint` is expected (since all integer
+types are subtypes of `mathint`).
+
+In order to convert from a supertype to a subtype, you must use an explicit
+cast.  In CVL 1, the syntax for casting to a subtype was `to_<subtype>(value)`,
+for example `to_uint256(x)`.
+
+In CVL 2, there are now two casting operators: `assert_<type>(value)` and
+`require_<type>(value)` (for example: `assert_uint8(x)` or `require_uint8(x)`).
+Each of these cases checks that the value is in range; the `assert` cast will
+report a counterexample if the value is out of range, while the `require` cast
+will ignore counterexamples where the cast value is out of range.
+
+```{todo}
+is it an error to use an explicit cast for an upcast?
+```
+
+```{todo}
+If you do not change this, you will see the following error:
+```
+
+### Modulo operator `%` returns negative values for negative inputs
+
+As in Solidity, if `n < 0` then `n % k == -(-n % k)`.
+
+### Support for `bytes1`...`bytes32`
+
+CVL 2 supports the types `bytes1`, `bytes2`, ..., `bytes32`, as in Solidity.
+Number literals must be explicitly cast to these types using `to_bytesN`; for
+example:
+
+```cvl
+bytes32 x = to_bytes32(0);
+```
+
+There is no way to convert between these types and integer types (except for
```

Comment:
the crowd will ask: does `to_mathint` not work on bytesN?

Reply:
bytes32 are described in solidity as "Fixed sized byte arrays" the fact that they are convertible to numbers is super weird imo.

Reply:
no need to repeat solidity's mistakes